### PR TITLE
Adding support for batch input in BERT Tokenizer with perf benchmark

### PR DESCRIPTION
**Pull request by parmeet**
This PR add support for batch tokenizer directly implemented at C++ kernel layer.

It also add benchmarking code for tokenizer. Following result is obtained by running tokenizer on 100000 samples.

```python
python benchmark/benchmark_bert_tokenizer.py --num-samples num_samples
```

**batch size 100**
Running TorchText BERT Tokenizer on non-batched input ... Total running time: 0.0014495829999998655
Running HF BERT Tokenizer (slow) on non-batched input ... Total running time: 0.01610425200000032
Running HF BERT Tokenizer (fast) on non-batched input ... Total running time: 0.0037178359999998634
Running TorchText BERT Tokenizer on batched input ... Total running time: 0.0007109650000001189
Running HF BERT Tokenizer (fast) on batched input ... Total running time: 0.0010531449999997555

**batch size 1000**
Running TorchText BERT Tokenizer on non-batched input ... Total running time: 0.021309904000000213
Running HF BERT Tokenizer (slow) on non-batched input ... Total running time: 0.383976283
Running HF BERT Tokenizer (fast) on non-batched input ... Total running time: 0.07770123400000006
Running TorchText BERT Tokenizer on batched input ... Total running time: 0.01940807400000022
Running HF BERT Tokenizer (fast) on batched input ... Total running time: 0.015243869999999937

**Observations**
*  ~15x on non-batched input compared to HF slow tokenizer (python based).
*  ~3x on non-batched input compared to HF fast tokenizer (Rust based)
*  torchtext is faster for batch processing up-to certain batch size but then as the batch size increases HF implementation becomes significantly faster (20-25%). This is likely because the kernel implementation of HF is faster compared to torchtext's but the overhead to call the binding for HF is higher compared to torchtext. 

Performance next steps:

- [ ] Perform analysis on the impact of batch size for batch processing.
- [ ] Identify and improve performance gaps in batch processing 
- [ ] Make use of timeit module in benchmark code

### CODE REVIEW

File: torchtext/transforms.py [hunk 1]
```diff
@@ -597,6 +597,23 @@ def _encode(self, text: str) -> List[str]:
...
 
+    @torch.jit.export
+    def _batch_encode(self, text: List[str]) -> List[List[str]]:
+        """Encode text into a list of tokens IDs
+
+        Args:
+            text: An input text string.
+
+        Returns:
+            A list of token ids represents each sub-word
```
Comment (Nayef211) [inline torchtext/transforms.py:608]: Can we update the docstring to reflect that we're operating on a list of strings and the output is a nested list of tokens

File: torchtext/transforms.py [hunk 2]
```diff
@@ -597,6 +597,23 @@ def _encode(self, text: str) -> List[str]:
...
+
+        For example:
+            --> "Hello world!" --> token ids: [707, 5927, 11, 707, 68]
```
Comment (Nayef211) [inline torchtext/transforms.py:611]: Also need to update the example

File: torchtext/transforms.py [hunk 3]
```diff
@@ -612,6 +629,21 @@ def _tokenize(self, text: str) -> List[str]:
...
+
+        Args:
+            text: An input text string.
+
+        Returns:
+            A list of tokens (sub-words)
+
+        For example:
+            --> "Hello World!": ["Hello", "World", "!"]
+        """
```
Comment (Nayef211) [inline torchtext/transforms.py:644]: Update docstring to reflect list input and nested list output

File: torchtext/csrc/register_pybindings.cpp [hunk 4]
```diff
@@ -220,6 +220,32 @@ PYBIND11_MODULE(_torchtext, m) {
       .def(py::init<const std::string, bool, c10::optional<bool>>())
       .def("encode", &BERTEncoder::Encode)
       .def("tokenize", &BERTEncoder::Tokenize)
+      .def(
+          "batch_encode",
+          [](const c10::intrusive_ptr<BERTEncoder>& self,
```
Comment (Nayef211) [inline torchtext/csrc/register_pybindings.cpp:225]: Is there a reason we have the operator implementation in the registration file as opposed to the `bert_tokenizer.cpp` file? I assume we'd like to keep registration code separate from the actual implementation code?

**Review by Nayef211 — COMMENTED** [review]
Left a couple of comments. In regards to the benchmark code, I wonder if we should try a really large value for `num_samples` so we can see how our batched implementation performs for very small and large batches. 

I also wonder if you've looked into using the `timeit` module which is meant for profiling and allows us to run a snippet of code several times to more accurately measure execution speed. I noticed that the `timeit` module has a `Timer` class built in. Could we potentially use this implementation instead of the one we have in `benchmark/utils.py`

File: torchtext/csrc/register_pybindings.cpp (hunk 4 above)
Comment (parmeet) [reply torchtext/csrc/register_pybindings.cpp:225]: That's a good catch. I moved the core implementation back into the class.

**Pull request merged**